cpp/include/partition_manager.hpp

```diff
@@ -0,0 +1,160 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/comms/comms.hpp>
+#include <raft/handle.hpp>
+
+#include <cassert>
+#include <cmath>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <vector>
+
+namespace cugraph {
+namespace partition_2d {
+template <typename from_t>
+std::string to_string(from_t const& value)
+{
+  std::stringstream ss;
+  ss << value;
+  return ss.str();
+}
+
+// default key-naming mechanism:
+//
+struct key_naming_t {
+  key_naming_t(int row_indx,
+               int col_indx,
+               std::string const& col_suffix = std::string("_col"),
+               std::string const& row_suffix = std::string("_row"),
+               std::string const& prefix     = std::string("partition"))
+    : col_suffix_(col_suffix),
+      row_suffix_(row_suffix),
+      prefix_(prefix),
+      name_(prefix_ + "_" + to_string(row_indx) + "_" + to_string(col_indx))
+  {
+  }
+
+  std::string col_name(void) const { return name_ + col_suffix_; }
+
+  std::string row_name(void) const { return name_ + row_suffix_; }
+
+ private:
+  std::string const col_suffix_;
+  std::string const row_suffix_;
+  std::string const prefix_;
+  std::string name_;
+};
+
+template <typename value_t>
+using matrix_t = std::vector<std::vector<value_t>>;
+
+using pair_comms_t =
+  std::pair<std::shared_ptr<raft::comms::comms_t>, std::shared_ptr<raft::comms::comms_t>>;
+
+enum class colors_2d_t : int { ROW = 0, COL = 1 };
+
+// class responsible for creating 2D partition sub-comms:
+//
+template <typename key_name_policy_t = key_naming_t, typename size_type = int>
+class partition_manager_t {
+ public:
+  partition_manager_t(raft::handle_t& handle, size_type p_row_size, size_type p_col_size)
+    : handle_(handle), p_row_size_(p_row_size), p_col_size_(p_col_size)
+  {
+    init_comms();
+  }
+
+  partition_manager_t(raft::handle_t const& handle, size_type p_size) : handle_(handle)
+  {
+    partition2d(p_size);
+    init_comms();
+  }
+
+  virtual ~partition_manager_t(void) {}
+
+  matrix_t<pair_comms_t> const& comms_matrix(void) const { return comms_set_; }
+
+ protected:
+  virtual void partition2d(size_type p_size)
+  {
+    auto sqr = static_cast<size_type>(std::sqrt(p_size));
+
+    // find divisor of p_size
+    // nearest to sqr;
+    //
+    p_row_size_ = nearest_divisor(sqr, p_size);
+    p_col_size_ = p_size / p_row_size_;
+
+    assert(p_row_size_ > 1 && p_col_size_ > 1);
+  }
+
+  void init_comms(void)
+  {
+    std::vector<pair_comms_t> empty_row(p_col_size_, std::make_pair(nullptr, nullptr));
+    comms_set_.assign(p_row_size_, empty_row);
+
+    raft::comms::comms_t const& communicator = handle_.get_comms();
+
+    for (size_type row = 0; row < p_row_size_; ++row)
+      for (size_type col = 0; col < p_col_size_; ++col) {
+        key_name_policy_t key{row, col};
```

Comment:
I am still reading this PR, but it seems like there is some misunderstanding about sub-communicators for 2D partitioning. Each client should have only two sub-communicators (and two keys).

One for row-wise communication, and another for column-wise communication.

Say, P = 4 (GPU0, GPU1, GPU2, and GPU3), P_row = 2, and P_col = 2.

i.e.

GPU0 GPU1
GPU2 GPU3

For row-wise communication, there should be one group having GPU0 & GPU1 (say row-group0) and another group having GPU2 and GPU3 (say row-group1).

For column-wise communication, there should be one group having GPU0 & GPU2 (say col-group0) and another group having GPU 1 & GPU3 (say col-group1).

So each GPU0, needs to run comm_split only twice to get row-group0 and col-group0.
In case of GPU 3, need to run comm_split twice to get row-group1 and col-group1.

No need to loop over p_row_size * p_col_size.

Reply:
See

https://github.com/rapidsai/cugraph/pull/1098/files#diff-75b979f478be47f71bd6933f474074cbR31
```
// FIXME: these should better be defined somewhere else.
std::string const comm_p_row_key = "comm_p_row";
std::string const comm_p_col_key = "comm_p_key";
```

and

https://github.com/rapidsai/cugraph/pull/1098/files#diff-51f3eab861b5285752c8047117a29f24R42
```
 auto &comm_p_row     = handle.get_subcomm(comm_p_row_key);
...
   comm_p_row.reduce(local_degrees.data(),
                      i == comm_p_col_rank ? degrees.data() : static_cast<edge_t *>(nullptr),
                      degrees.size(),
                      raft::comms::op_t::SUM,
                      comm_p_col_rank,
                      handle.get_stream());
```

For the ```handle.get_subcomm(comm_p_row_key);``` to work, ```set_subcomm``` should be called somewhere before this, and the names for the row-wise and column-wise communication keys should be defined (I temporarily defined the key names in graph_view.hpp but I think there will be a better place).

And this should be called when initializing our default_handle in python initialization. @Iroy30 may have a suggestion about the best place to call a function that initializes sub-communicators (two per GPU) and adds those to the default handle.

Reply:
And row-group0 and row-group1 can use the same key name and col-group0 & col-group1 can use the same key. On GPU0, only row-group0 and col-group0 are visible, so no need to worry about row-group1.

Reply:
A reading material:

https://mpitutorial.com/tutorials/introduction-to-groups-and-communicators

NCCL Comm split (and RAFT communicator split) is pretty similar to MPI Comm split.

Reply:
Row/column key-naming aside, I think I've fixed the `subcomm` creation from a loop (one dedicated worker approach) to a row-and-column (hence two subcomms) per-worker approach.

As for the key-naming mechanism, is not clear to me how that should be fixed, yet. Taking this offline.